File: hipe-doc/ghc/Bench.hs
{-# language Strict, LambdaCase, BlockArguments, OverloadedStrings #-}

-- ghc -O1 -threaded -rtsopts -fforce-recomp -fworker-wrapper-cbv Bench.hs

import Data.Time.Clock
import Control.Monad
import System.IO
import Data.String

timed :: String -> Int -> (a -> IO b) -> a -> IO ()
timed msg times act a = do
  buffering <- hGetBuffering stdout
  hSetBuffering stdout NoBuffering
  t0 <- getCurrentTime
  let dot _ = putStr ""; {-# noinline dot #-}
  replicateM_ times $ dot =<< act a
  t1 <- getCurrentTime
  let pad = replicate (20 - length msg) ' '
  putStrLn $ msg ++ ":" ++ pad ++ show (diffUTCTime t1 t0 / fromIntegral times)
  hSetBuffering stdout buffering
{-# noinline timed #-}

data RTm
  = RVar String
  | RApp RTm RTm
  | RLam String RTm

data Tm
  = Var Int
  | App Tm Tm
  | Lam String Tm
  deriving Show

elabVar :: [String] -> String -> Int
elabVar ns x = case ns of
  x':ns | x == x' -> 0
        | True    -> elabVar ns x + 1
  _ -> undefined

elab' :: [String] -> RTm -> Tm
elab' ns = \case
  RVar x   -> Var (elabVar ns x)
  RApp t u -> App (elab' ns t) (elab' ns u)
  RLam x t -> Lam x (elab' (x:ns) t)

elab = elab' []

data Env = Nil | Cons Val Env

data Val =
    VVar Int
  | VApp Val Val
  | VLam String Env Tm

vvar e (x :: Int) = case (e, x) of
  (Cons v _, 0) -> v
  (Cons _ e, x) -> vvar e (x - 1)
  _             -> undefined

eval e = \case
  Var x -> vvar e x
  App t u -> case eval e t of
    VLam _ e' t -> eval (Cons (eval e u) e') t
    t           -> VApp t (eval e u)
  Lam x t -> VLam x e t

conv (l :: Int) t t' = case (t, t') of
  (VVar x, VVar x') -> x == x'
  (VApp t u, VApp t' u') -> conv l t t' && conv l u u'
  (VLam x e t, VLam x' e' t') ->
    let v = VVar l in
    conv (l + 1) (eval (Cons v e) t) (eval (Cons v e') t')
  _ -> False

quote (l :: Int) = \case
  VVar x -> Var (l - x - 1)
  VApp t u -> App (quote l t) (quote l u)
  VLam x e t -> Lam x (quote (l + 1) (eval (Cons (VVar l) e) t))

eval0 = eval Nil
conv0 = conv 0
quote0 = quote 0
nf0 = quote0 . eval0

infixl 8 $$
($$) :: RTm -> RTm -> RTm
($$) = RApp
{-# inline ($$) #-}

instance IsString RTm where
  fromString = RVar

let_ :: String -> RTm -> RTm -> RTm
let_ x t u = RLam x u $$ t
{-# inline let_ #-}

prog :: RTm
prog =
  let_ "zero" (RLam "s" $ RLam "z" "z") $
  let_ "suc" (RLam "n" $ RLam "s" $ RLam "z" $ "s" $$ ("n" $$ "s" $$ "z")) $
  let_ "n5" ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ "zero"))))) $
  let_ "add" (RLam "n" $ RLam "m" $ RLam "s" $ RLam "z" $ "n" $$ "s" $$ ("m" $$ "s" $$ "z")) $
  let_ "n10" ("add" $$ "n5" $$ "n5") $
  let_ "n15" ("add" $$ "n5" $$ "n10") $
  let_ "n20" ("add" $$ "n5" $$ "n15") $
  let_ "leaf" (RLam "n" $ RLam "l" $ "l") $
  let_ "node" (RLam "t1" $ RLam "t2" $ RLam "n" $ RLam "l" $ "n" $$ ("t1" $$ "n" $$ "l")
                                                                 $$ ("t2" $$ "n" $$ "l")) $
  let_ "mktree" (RLam "n" $ "n" $$ (RLam "t" $ "node" $$ "t" $$ "t") $$ "leaf") $
  "mktree" $$ "n20"

prog2 :: RTm
prog2 =
  let_ "zero" (RLam "s" $ RLam "z" "z") $
  let_ "suc" (RLam "n" $ RLam "s" $ RLam "z" $ "s" $$ ("n" $$ "s" $$ "z")) $
  let_ "n5" ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ "zero"))))) $
  let_ "add" (RLam "n" $ RLam "m" $ RLam "s" $ RLam "z" $ "n" $$ "s" $$ ("m" $$ "s" $$ "z")) $
  let_ "n10" ("add" $$ "n5" $$ "n5") $
  let_ "n15" ("add" $$ "n5" $$ "n10") $
  let_ "n20" ("add" $$ "n5" $$ "n15") $
  let_ "leaf" (RLam "n" $ RLam "l" $ "l") $
  let_ "node" (RLam "t1" $ RLam "t2" $ RLam "n" $ RLam "l" $ "n" $$ ("t1" $$ "n" $$ "l")
                                                                 $$ ("t2" $$ "n" $$ "l")) $
  let_ "mktree" (RLam "n" $ "n" $$ (RLam "t" $ "node" $$ "t" $$ "t") $$ "leaf") $
  "mktree" $$ "n20" $$ (RLam "_" $ RLam "_" $ "zero") $$ "zero"

prog3 :: RTm
prog3 =
  let_ "zero" (RLam "s" $ RLam "z" "z") $
  let_ "suc" (RLam "n" $ RLam "s" $ RLam "z" $ "s" $$ ("n" $$ "s" $$ "z")) $
  let_ "n5" ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ "zero"))))) $
  let_ "add" (RLam "n" $ RLam "m" $ RLam "s" $ RLam "z" $ "n" $$ "s" $$ ("m" $$ "s" $$ "z")) $
  let_ "n10" ("add" $$ "n5" $$ "n5") $
  let_ "n15" ("add" $$ "n5" $$ "n10") $
  let_ "n20" ("add" $$ "n5" $$ "n15") $
  let_ "n25" ("add" $$ "n5" $$ "n20") $
  let_ "mktree" (RLam "n" $ RLam "node" $ RLam "l" $
                 "n" $$ (RLam "x" $ "node" $$ "x" $$ "x") $$ "l") $
  "mktree" $$ "n20"

prog4 :: RTm
prog4 =
  let_ "zero" (RLam "s" $ RLam "z" "z") $
  let_ "suc" (RLam "n" $ RLam "s" $ RLam "z" $ "s" $$ ("n" $$ "s" $$ "z")) $
  let_ "n5" ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ ("suc" $$ "zero"))))) $
  let_ "add" (RLam "n" $ RLam "m" $ RLam "s" $ RLam "z" $ "n" $$ "s" $$ ("m" $$ "s" $$ "z")) $
  let_ "n10" ("add" $$ "n5" $$ "n5") $
  let_ "n15" ("add" $$ "n5" $$ "n10") $
  let_ "n20" ("add" $$ "n5" $$ "n15") $
  let_ "n25" ("add" $$ "n5" $$ "n20") $
  let_ "mktree" (RLam "n" $ RLam "node" $ RLam "l" $
                  "n" $$ (RLam "x" $ "node" $$ "x" $$ "x") $$ "l") $
  "mktree" $$ "n20" $$ (RLam "_" $ RLam "_" $ "zero") $$ "zero"

run = show . nf0 . elab

{-# noinline force #-}
force t = case nf0 (elab t) of
  Lam{} -> 0 :: Int
  _     -> 1 :: Int

data Tree = Node Tree Tree | Leaf Int
  deriving Show

maptree :: Tree -> Tree
maptree = \case
  Leaf n   -> Leaf (n + 100)
  Node l r -> Node (maptree l) (maptree r)

mktree :: Int -> Tree
mktree n = go n n where
  go 0 _ = Leaf 0
  go _ 0 = Leaf 0
  go n m = Node (go (n - 1) (m - 1)) (go (m - 1) (n - 1))

iter = 30 :: Int

main = do
  timed "Tree NF" iter (pure . force) prog
  timed "Tree Conv" iter (\t -> let v = eval0 (elab t) in pure $ conv0 v v) prog
  timed "Tree force" iter (pure . force) prog2
  timed "Tree NF share" iter (pure . force) prog3
  timed "Tree Conv share" iter (\t -> let v = eval0 (elab t) in pure $ conv0 v v) prog3
  timed "Tree force share" iter (pure . force) prog4

  timed "Maptree 1/2" iter (\n -> pure $ seq (maptree (mktree n)) ()) 20
  timed "Maptree 2/3" iter (\n -> pure $ seq (maptree(maptree (mktree n))) ()) 20
  timed "Maptree 3/4" iter (\n -> pure $ seq (maptree(maptree(maptree (mktree n)))) ()) 20
  timed "Maptree 4/5" iter (\n -> pure $ seq (maptree(maptree(maptree(maptree(mktree n))))) ()) 20

File: hipe-doc/ocaml/bench.ml


type rtm
  = RVar of string
  | RApp of rtm * rtm
  | RLam of string * rtm

type tm
  = Var of int
  | App of tm * tm
  | Lam of string * tm

let rec show(t : tm) =
  match t with
  | Var(x)    -> "Var(" ^ string_of_int x ^ ")"
  | App(t, u) -> "App(" ^ show t ^ ", " ^ show u ^ ")"
  | Lam(x, t) -> "Lam(" ^ x ^ ", " ^ show t ^ ")"

let undef() = raise (Failure "undefined")

let rec elabVar (ns : string list) (x : string) =
  match ns with
  | x' :: ns -> if x == x' then 0 else elabVar ns x + 1
  | _        -> undef()

let rec elab' (ns : string list) = function
  | RVar(x)    -> Var(elabVar ns x)
  | RApp(t, u) -> App(elab' ns t, elab' ns u)
  | RLam(x, t) -> Lam(x, elab' (x::ns) t)

let elab = elab' []

type value
  = VVar of int
  | VApp of value * value
  | VLam of string * env * tm

and env = value list

let rec vvar (e : env) (x : int) = match (e, x) with
  | v :: _, 0 -> v
  | _ :: e, x -> vvar e (x - 1)
  | _         -> undef()

let rec eval e = function
  | Var(x)    -> vvar e x
  | App(t, u) -> (match eval e t with
                   | VLam(_, e', t) -> eval (eval e u :: e') t
                   | t              -> VApp(t, eval e u))
  | Lam(x, t) -> VLam(x, e, t)

let rec conv l t t' = match (t, t') with
  | VVar(x), VVar(x') -> x == x'
  | VApp(t, u), VApp(t', u') -> conv l t t' && conv l u u'
  | VLam(x, e, t), VLam(x', e', t') ->
     let v = VVar(l) in
     conv (l + 1) (eval (v :: e) t) (eval (v :: e') t')
  | _ -> false

let rec quote (l : int) = function
  | VVar(x)       -> Var(l - x - 1)
  | VApp(t, u)    -> App(quote l t, quote l u)
  | VLam(x, e, t) -> Lam(x, quote (l + 1) (eval (VVar(l)::e) t))

let eval0 = eval []
let conv0 = conv 0
let quote0 = quote 0
let nf0 x = quote0 (eval0 x)

let let'(x, t, u)     = RApp(RLam(x, u), t) [@@inline]
let rapp2(t, u, v)    = RApp(RApp(t, u), v) [@@inline]
let rapp3(t, u, v, w) = RApp(RApp(RApp(t, u), v), w) [@@inline]

let vx = RVar("x")
let vy = RVar("y")
let vz = RVar("z")
let vs = RVar("s")
let vt = RVar("t")
let vn = RVar("n")
let vm = RVar("m")
let vc = RVar("c")
let vl = RVar("l")
let vr = RVar("r")

let prog =
  let'("zero", RLam("s", RLam("z", vz)),
  let'("suc", RLam("n", RLam("s", RLam("z", RApp(vs, rapp2(vn, vs, vz))))),
  let'("five", RApp(RVar("suc"), RApp(RVar("suc")
              , RApp(RVar("suc"), RApp(RVar("suc"), RApp(RVar("suc"), RVar("zero")))))),
  let'("add", RLam("n", RLam("m", RLam("s", RLam("z", rapp2(vn, vs, rapp2(vm, vs, vz)))))),
  let'("ten", rapp2(RVar("add"), RVar("five"), RVar("five")),
  let'("n15", rapp2(RVar("add"), RVar("five"), RVar("ten")),
  let'("n20", rapp2(RVar("add"), RVar("five"), RVar("n15")),
  let'("leaf", RLam("n", RLam("l", RVar("l"))),
  let'("node", RLam("t1", RLam("t2", RLam("n", RLam("l",
              rapp2(vn, rapp2(RVar("t1"), vn, vl), rapp2(RVar("t2"), vn, vl)))))),
  let'("mktree", RLam("n", rapp2(vn, (RLam("t", rapp2(RVar("node"), vt, vt))), RVar("leaf"))),
  RApp(RVar("mktree"), (RVar("n20")))
  ))))))))))

let prog2 =
  let'("zero", RLam("s", RLam("z", vz)),
  let'("suc", RLam("n", RLam("s", RLam("z", RApp(vs, rapp2(vn, vs, vz))))),
  let'("five", RApp(RVar("suc"), RApp(RVar("suc")
              , RApp(RVar("suc"), RApp(RVar("suc"), RApp(RVar("suc"), RVar("zero")))))),
  let'("add", RLam("n", RLam("m", RLam("s", RLam("z", rapp2(vn, vs, rapp2(vm, vs, vz)))))),
  let'("ten", rapp2(RVar("add"), RVar("five"), RVar("five")),
  let'("n15", rapp2(RVar("add"), RVar("five"), RVar("ten")),
  let'("n20", rapp2(RVar("add"), RVar("five"), RVar("n15")),
  let'("leaf", RLam("n", RLam("l", RVar("l"))),
  let'("node", RLam("t1", RLam("t2", RLam("n", RLam("l",
              rapp2(vn, rapp2(RVar("t1"), vn, vl), rapp2(RVar("t2"), vn, vl)))))),
  let'("mktree", RLam("n", rapp2(vn, (RLam("t", rapp2(RVar("node"), vt, vt))), RVar("leaf"))),
  rapp2(RApp(RVar("mktree"), (RVar("n20"))), RLam("_", RLam("_", RVar("leaf"))), RVar("leaf"))
  ))))))))))

let prog3 =
  let'("zero", RLam("s", RLam("z", vz)),
  let'("suc", RLam("n", RLam("s", RLam("z", RApp(vs, rapp2(vn, vs, vz))))),
  let'("five", RApp(RVar("suc"), RApp(RVar("suc")
              , RApp(RVar("suc"), RApp(RVar("suc"), RApp(RVar("suc"), RVar("zero")))))),
  let'("add", RLam("n", RLam("m", RLam("s", RLam("z", rapp2(vn, vs, rapp2(vm, vs, vz)))))),
  let'("ten", rapp2(RVar("add"), RVar("five"), RVar("five")),
  let'("n15", rapp2(RVar("add"), RVar("five"), RVar("ten")),
  let'("n20", rapp2(RVar("add"), RVar("five"), RVar("n15")),
  let'("n25", rapp2(RVar("add"), RVar("five"), RVar("n20")),
  let'("mktree", RLam("n", RLam("node", RLam("l", rapp2(RVar("n")
                     , RLam("x", rapp2(RVar("node"), RVar("x"), RVar("x")))
                     , RVar("l"))))),
  RApp(RVar("mktree"), (RVar("n20")))
  )))))))))

let prog4 =
  let'("zero", RLam("s", RLam("z", vz)),
  let'("suc", RLam("n", RLam("s", RLam("z", RApp(vs, rapp2(vn, vs, vz))))),
  let'("five", RApp(RVar("suc"), RApp(RVar("suc")
              , RApp(RVar("suc"), RApp(RVar("suc"), RApp(RVar("suc"), RVar("zero")))))),
  let'("add", RLam("n", RLam("m", RLam("s", RLam("z", rapp2(vn, vs, rapp2(vm, vs, vz)))))),
  let'("ten", rapp2(RVar("add"), RVar("five"), RVar("five")),
  let'("n15", rapp2(RVar("add"), RVar("five"), RVar("ten")),
  let'("n20", rapp2(RVar("add"), RVar("five"), RVar("n15")),
  let'("n25", rapp2(RVar("add"), RVar("five"), RVar("n20")),
  let'("mktree", RLam("n", RLam("node", RLam("l", rapp2(RVar("n")
                     , RLam("x", rapp2(RVar("node"), RVar("x"), RVar("x")))
                     , RVar("l"))))),
  rapp2(RApp(RVar("mktree"), (RVar("n20"))), RLam("_", RLam("_", RVar("zero"))), RVar("zero"))
    )))))))))

let run x = show (nf0 (elab x))

let force t : int = match nf0 (elab t) with
  | Lam(_) -> 0
  | _      -> 1

type tree
  = Node of tree * tree
  | Leaf of int

let rec maptree = function
  | Leaf(n)    -> Leaf(n + 100)
  | Node(l, r) -> Node(maptree l, maptree r)

let rec mktree (n : int) =
  let rec go n n = match (n, n) with
    | 0, _ -> Leaf(0)
    | _, 0 -> Leaf(0)
    | n, m -> Node(go (n - 1) (m - 1), go (m - 1) (n - 1)) in
  go n n

let timed (msg : string) (times : int)(act : 'a -> 'b) (start : 'a) =
  let t = Sys.time() in
  for i = 1 to times do
    let _ = act start in
    ()
  done;
  let t' = Sys.time() in
  let pad = String.make (20 - String.length msg) ' ' in
  Printf.printf "%s:%s%f\n%!" msg pad ((t' -. t) /. (float_of_int times))
[@@noinline]

let iter : int = 30

let _ = timed "Tree NF" iter force prog
let _ = timed "Tree Conv" iter (fun t -> let v = eval0 (elab t) in conv0 v v) prog
let _ = timed "Tree force" iter force prog2
let _ = timed "Tree NF share" iter force prog3
let _ = timed "Tree Conv share" iter (fun t -> let v = eval0 (elab t) in conv0 v v) prog3
let _ = timed "Tree force share" iter force prog4
let _ = timed "Maptree 1/2" iter (fun n -> let _ = maptree(mktree n) in ()) 20
let _ = timed "Maptree 2/3" iter (fun n -> let _ = maptree(maptree (mktree n)) in ()) 20
let _ = timed "Maptree 3/4" iter (fun n -> let _ = maptree(maptree(maptree (mktree n))) in ()) 20
let _ = timed "Maptree 4/5" iter (fun n -> let _ = maptree(maptree(maptree(maptree(mktree n)))) in ()) 20

(* export OCAMLRUNPARAM="v=0x400,s=4000000" *)
(* /usr/bin/time -v ./bench  *)
